### PR TITLE
EDSC-3966: Single Harmony service selection box

### PR DESCRIPTION
**Pull request by rushgeo**
# Overview

### What is the feature?

A single selection box is shown for available Harmony transformation services rather than a box for each service. This also improves the access method selection UX through simplified language, additional icons, and tooltips.

### What is the Solution?

Users select the Harmony type of access method in step 1 of the access method panel, then the specific Harmony method and its options in step 2.

No changes to the URL routes were necessary, so bookmarked URLs or saved projects with a selected access method should continue to work.

### Example screenshots

Using the collection `C1996881146-POCLOUD`:

| Before |
| ------------- |
| access methods: before  

| After (Service List) | After (Service Selected) |
| ------------- | ------------- |
| access methods: after, selection box | access methods: after  |

### What areas of the application does this impact?

Selecting access methods and options before downloading selected data.

# Testing

After running this branch locally:

- Open http://localhost:8080/projects?p=C1996881146-POCLOUD!C1996881146-POCLOUD
- Verify access method details appear as a tooltip over the (?) icons on the right side of each item in Step 1
- Select `Customize with OPeNDAP` and `Download all data` and verify those methods continue to work as before
- Select `Customize with Harmony`
- Click the service selction box in Step 2
- Verify four Harmony methods appear in the selection popup, and you can scroll the popup to view all four
- Verify each of the Harmony methods in the selection popup have icons in their top right corner with tooltips on mouseover that indicate their customization capabilities
- Select a Harmony method from the selection popup
- Verify the service name appears in the selection box
- Verify its name, description, and a link to the Harmony documentation appears below the selection box
- Verify that the appropriate customization options appear below the link to the documentation

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings

### CODE REVIEW

File: static/src/js/components/AccessMethod/Select.css [hunk 1]
```diff
@@ -0,0 +1,101 @@
+@import '@radix-ui/colors/black-alpha.css';
+
+/* reset */
+button {
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/Select.css:4]: So in the select we are trying to unset all the styling for buttons in this component but, I'm not sure why? We should be able to accomplish this with `classNames`

File: static/src/js/util/metaIcon.js [hunk 2]
```diff
@@ -0,0 +1,214 @@
+import React from 'react'
```
Comment (eudoroolivares2016) [inline static/src/js/util/metaIcon.js:1]: We should add a test for this component.

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 3]
```diff
@@ -204,6 +232,52 @@ export class AccessMethod extends Component {
...
+      })
+
+      this.handleAccessMethodSelection(event)
+    }
+  }
+
+  harmonyMethodsMapper(methods, selected) {
+    return (
+      <div id="harmony_methods">
+        {methods.map((radio) => this.renderHarmonySelectItem(radio, selected))}
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:248]:
```suggestion
        {methods.map((radio) => this.createHarmonySelectItem(radio, selected))}
```
Just nitpick I'd like to just obfuscate the word render from this

File: static/src/js/util/metaIcon.js [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+ * Used by CollectionResultsItem and AccessMethodRadio.
+ * @param {boolean} hasSpatialSubsetting
+ * @param {boolean} hasVariables
+ * @param {boolean} hasTransforms
+ * @param {boolean} hasFormats
+ * @param {boolean} hasTemporalSubsetting
+ * @param {boolean} hasCombine
+ * @param {boolean} forAccessMethodRadio - indicates usage for AccessMethodRadio
+ */
+export function getCustomizeIcons(
```
Comment (eudoroolivares2016) [inline static/src/js/util/metaIcon.js:27]: I think if you want this as a util we should rename this file to `getCustomizeIcons`. We already have a `static/src/js/components/MetaIcon/MetaIcon.js`

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 5]
```diff
@@ -204,6 +232,52 @@ export class AccessMethod extends Component {
     })
   }
 
+  handleHarmonySelection(event, harmonyMethods) {
+    if (event !== '') {
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:236]: Can this be 
```suggestion
    if (event) {
```
?

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 6]
```diff
@@ -204,6 +232,52 @@ export class AccessMethod extends Component {
...
+      this.setState({
+        selectValue: harmonyMethods.find(({ methodKey }) => methodKey === event).name
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:238]: a bit confused here can't we just destructure that value `methodKey` from the harmonyMethods then pull out the name from that?

```suggestion
        selectValue: harmonyMethods.find(({ methodKey }) => methodKey === event).name
```

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 7]
```diff
@@ -300,20 +474,36 @@ export class AccessMethod extends Component {
...
-          description = 'Select options like variables, transformations, and output formats for in-region cloud access.'
-          details = `The requested data will be processed using the ${name} service and stored in the cloud for analysis.`
+          title = name
+          description = descriptionFromMetadata
+          hasHarmony = true
+          hasSpatialSubsetting = hasShapefileSubsetting || hasBBoxSubsetting
+          hasFormats = supportedOutputFormats.length > 1
+          hasProjections = supportedOutputProjections.length > 1
+          /**
+           * CMR defines has-transforms as including interpolation or multiple projections, but
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:493]: Lets try to rephrase this as a TODO with some action like: `TODO: Add support for interpolation`

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 8]
```diff
@@ -530,6 +720,26 @@ export class AccessMethod extends Component {
                     </ProjectPanelSection>
                   )
                 }
+                {
+                  // Show Harmony method description
+                  harmonyTypeSelected && (
+                    <>
+                      <p />
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:727]: Lets try to simplify the number of HTML elements we're using here I think some of them are unnecessary. I think this would be wrapped by a `div` instead of a fragment

File: static/src/js/components/EDSCIcon/EDSCIcon.scss [hunk 9]
```diff
@@ -3,4 +3,12 @@
...
   }
+  &--details {
+    color: $color__black--300;
+    height: 100%;
+  }
 }
+
+.rightside {
+  margin-left: auto;
+}
```
Comment (eudoroolivares2016) [inline static/src/js/components/EDSCIcon/EDSCIcon.scss:14]: newline

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 10]
```diff
@@ -231,6 +305,88 @@ export class AccessMethod extends Component {
...
+              </ScrollArea.Scrollbar>
+              <Select.Arrow />
+            </ScrollArea.Root>
+          </Select.Content>
+        </Select.Portal>
+      </Select.Root>
+    )
+  }
+
+  renderStep1(hasHarmony, radioList, collectionId, selectedAccessMethod) {
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:352]: This one I do think we should really rename to something a little more clear to what its doing

File: static/src/js/components/CollectionResults/CollectionResultsItem.js [hunk 11]
```diff
@@ -201,176 +196,14 @@ export const CollectionResultsItem = forwardRef(({
...
-                                    Combine
-                                  </li>
-                                )
-                              }
-                            </ul>
-                          </>
-                        )
-                      }
-                    />
+                  getCustomizeIcons(
```
Comment (bnp26) [inline static/src/js/components/CollectionResults/CollectionResultsItem.js:199]: You wanna rename this getCustomizedIcons or getCustomIcons?

**Comment by rushgeo** [comment]
I'm getting intermittent failures of a random sample of the TemporalSelectionDropdown tests on my slow laptop, but they just passed in the GH Action. Locally, four or five will fail if I run `npm test`, but pass if I test just `TemporalSelectionDropdown.test.js` or `AdminRetrievalsMetrics.test.js`.

We may need to increase the timeouts on the `await`s to reduce noise in local dev and reduce the risk of future intermittent test failures.

File: static/src/js/components/AccessMethod/Select.css (hunk 1 above)
Comment (rushgeo) [reply static/src/js/components/AccessMethod/Select.css]: This was a remnant of the sample CSS from the Radix component's docs page; I removed it and some other unused styles when converting the CSS to SCSS.

File: static/src/js/components/AccessMethod/AccessMethod.js (hunk 10 above)
Comment (rushgeo) [reply static/src/js/components/AccessMethod/AccessMethod.js]: I renamed it to `getAccessMethodTypes()`, since that's the content of the Step 1 part of the access methods panel. How does that sound?

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 12]
```diff
@@ -75,10 +78,14 @@ export class AccessMethod extends Component {
       )
     } = selectedMethod || {}
 
+    const harmonyTypeSelected = selectedAccessMethod ? selectedAccessMethod.startsWith('harmony') : false
```
[the remark this answers is not in the document]
Comment (rushgeo) [reply static/src/js/components/AccessMethod/AccessMethod.js]: Good catch! Latest commit removes the redundancy and standardizes on isHarmony

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 13]
```diff
@@ -474,40 +684,43 @@ export class AccessMethod extends Component {
...
         <ProjectPanelSection
-          heading="Select a data access method"
-          intro="The selected access method will determine which customization and output options are available."
+          heading="Choose how you want to download your data"
           step={1}
         >
           <div className="access-method__radio-list">
             {
-              radioList.length === 0
+              radioList.length === 0 && hasHarmony === false
```
Comment (rushgeo) [inline static/src/js/components/AccessMethod/AccessMethod.js:697]: remove hasHarmony from check

File: static/src/js/components/AccessMethod/Select.scss [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  border-radius: 3px;
+  display: flex;
+  align-items: center;
+  height: auto;
+  position: relative;
+  user-select: none;
+}
+.SelectItem[data-highlighted] {
+  outline: none;
+  background-color: $color__black--200;
```
Comment (rushgeo) [inline static/src/js/components/AccessMethod/Select.scss:58]: make the same color as a hovered collection results item

File: static/src/js/components/AccessMethod/__tests__/AccessMethod.test.js [hunk 15]
```diff
@@ -408,13 +412,131 @@ describe('AccessMethod component', () => {
...
+              supportsConcatenation: true
+            }
+          },
+          selectedAccessMethod: 'harmony0',
+          metadata: {
+            conceptId: collectionId
+          }
+        })
+
+        expect(screen.getByText('test description').toBeInTheDocument)
```
Comment (rushgeo) [inline static/src/js/components/AccessMethod/__tests__/AccessMethod.test.js:528]: should this be a function, and outside of expect?

File: static/src/js/components/AccessMethod/AccessMethod.js (hunk 13 above)
Comment (rushgeo) [reply static/src/js/components/AccessMethod/AccessMethod.js:697]: Nevermind - `radiolist` only contains the non-Harmony methods now, so we need to check for it being empty as well as the Harmony method type being selected.

File: static/src/js/util/metaIcon.js (hunk 2 above)
Comment (rushgeo) [reply static/src/js/util/metaIcon.js]: This isn't a new component, it's just relocated code from CollectionResultsItem.js to its own file in 6d42317 since it is now also being called from AccessMethodRadio.js. It's still being tested through the those other two scripts.

File: static/src/js/components/CollectionResults/CollectionResultsItem.js (hunk 11 above)
Comment (rushgeo) [reply static/src/js/components/CollectionResults/CollectionResultsItem.js:199]: This is for getting the icons that show how a download is customizable (e.g. concatenation, spatial subsetting, temporal, etc.)
I'm not sure `getCustomizedIcons` or `getCustomIcons` is more clear than `getCustomizeIcons`. Maybe `getCustomizableIcons`? But that could be confusing because just looking at the name someone might think the icons are customizable, but it's the icons indicating how the data is customizable.

File: static/src/js/util/getCustomizeIcons.js
Comment (macrouch) [inline static/src/js/util/getCustomizeIcons.js:1]: This should be a component, not a util function.

We don't use `export function ...` syntax, it should be a `const`

When you have 7 parameters do not use 7 ordered parameters, that is a maintenance nightmare. Make it an object where each parameter has a useful name. This is irrelevant because it should be a component and they should be props.

@eudoroolivares2016 comment about tests brought me to this PR. This component absolutely should be tested individually, all out components should have unit tests

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 16]
```diff
@@ -75,10 +78,14 @@ export class AccessMethod extends Component {
...
 
+    const harmonyTypeSelected = selectedAccessMethod ? selectedAccessMethod.startsWith('harmony') : false
+
     this.state = {
       enableTemporalSubsetting,
       enableSpatialSubsetting,
-      enableConcatenateDownload
+      enableConcatenateDownload,
+      harmonyTypeSelected,
+      selectValue: ''
```
[the remark this answers is not in the document]
Comment (rushgeo) [reply static/src/js/components/AccessMethod/AccessMethod.js]: Renamed to `selectedHarmonyMethodName`

File: static/src/js/util/getCustomizeIcons.js
Comment (rushgeo) [reply static/src/js/util/getCustomizeIcons.js:1]: You didn't reply in this thread so I wanted to make sure you saw it: https://github.com/nasa/earthdata-search/pull/1726#discussion_r1534419781

I didn't write the function, I only moved it. If it needs to be rewritten as a component regardless, I can do that, but I wanted the full context brought together.

File: static/src/js/components/CollectionResults/CollectionResultsItem.js (hunk 11 above)
Comment (bnp26) [reply static/src/js/components/CollectionResults/CollectionResultsItem.js:199]: Whichever you think is right is fine. @eudoroolivares2016 if you could chime in I'd appreciate another brain on this,

File: static/src/js/util/getCustomizeIcons.js
Comment (rushgeo) [reply static/src/js/util/getCustomizeIcons.js:1]: Oh wait, I *did* write the function around the existing code. Sorry, it's been a couple weeks. Yeah it will be better as a component.

Comment (rushgeo) [reply static/src/js/util/getCustomizeIcons.js:1]: Converted to a component in [b7d71c8](https://github.com/nasa/earthdata-search/pull/1726/commits/b7d71c8f4b762923e5c394aa1356733c5548e779), tests coming soon.

File: static/src/js/components/CollectionResults/CollectionResultsItem.js [hunk 17]
```diff
@@ -5,22 +5,17 @@ import { OverlayTrigger, Tooltip } from 'react-bootstrap'
...
-  FaPlus,
-  FaSlidersH,
-  FaTags
+  FaPlus
 } from 'react-icons/fa'
 
 import { collectionMetadataPropType } from '../../util/propTypes/collectionMetadata'
 import { commafy } from '../../util/commafy'
 import { getApplicationConfig } from '../../../../../sharedUtils/config'
+import CustomizableIcons from '../CustomizableIcons/CustomizableIcons'
```
Comment (eudoroolivares2016) [inline static/src/js/components/CollectionResults/CollectionResultsItem.js:18]: Move this down with the other imports for components

File: static/src/js/components/AccessMethod/AccessMethod.js [hunk 18]
```diff
@@ -75,10 +80,23 @@ export class AccessMethod extends Component {
       )
     } = selectedMethod || {}
 
+    const isHarmony = selectedAccessMethod
+      ? selectedAccessMethod.startsWith('harmony')
+      : false
+
+    let selectedHarmonyMethodName = ''
+    if (selectedAccessMethod
+        && selectedAccessMethod.startsWith('harmony')
```
Comment (eudoroolivares2016) [inline static/src/js/components/AccessMethod/AccessMethod.js:89]:
```suggestion
        && isHarmony
```
isn't this already being computed above?

File: static/src/js/components/FormFields/AccessMethodRadio/AccessMethodRadio.js [hunk 19]
```diff
@@ -19,22 +19,30 @@ export const AccessMethodRadio = ({
...
-  const labelClassName = classNames([
+  const labelClasses = [
     'access-method-radio',
     {
       'access-method-radio--is-selected': checked
     }
-  ])
+  ]
+  if (isHarmony) {
+    labelClasses.push('access-method-radio--isHarmony')
```
Comment (trevorlang) [inline static/src/js/components/FormFields/AccessMethodRadio/AccessMethodRadio.js:33]: `--isHarmony` should be `--is-harmony` for consistency

File: package.json [hunk 20]
```diff
@@ -102,6 +102,9 @@
     "@edsc/geo-utils": "^1.0.4",
     "@edsc/smart-handoffs": "^1.0.5",
     "@edsc/timeline": "^1.1.7",
+    "@radix-ui/colors": "3.0.0",
```
Comment (trevorlang) [inline package.json:105]: Why was `colors` added? I am not seeing the radix colors used anywhere. Which is good. We should be using the colors available in the scss variables when possible.

File: static/src/js/components/AccessMethod/Select.scss [hunk 21]
```diff
@@ -0,0 +1,89 @@
+$trigger-width: 400px;
+
+.SelectTrigger {
+  width: $trigger-width;
+  border-radius: 4px;
+  border: 1px solid $color__black--200;
+  padding: 0 15px;
+  font-size: 13px;
```
Comment (trevorlang) [inline static/src/js/components/AccessMethod/Select.scss:8]: We should be using `rem`s for all fixed sizing. That includes text, margins, padding, etc. We rely on that behavior to do some nice scaling depending on screen size. You can get the proper size in rems by dividing the px value by 16. For example `13px / 16 = 0.8125rem`. For margins and padding, I like to calculate the variable based of the `$spacer` var (when I can remember 😅) so thats all consistent.

File: static/src/js/components/AccessMethod/Select.scss [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+  user-select: none;
+  padding: 2px;
+  &[data-orientation='vertical'] {
+    width: var(--scrollbar-size);
+  }
+}
+
+.ScrollAreaThumb {
+  background: $color__black--300;
+  border-radius: 3px;
```
Comment (trevorlang) [inline static/src/js/components/AccessMethod/Select.scss:86]: We have some `$border-radius` vars that could be used here

**Review by trevorlang — CHANGES_REQUESTED** [review]

**Comment by trevorlang** [comment]
@rushgeo, requested some changes on this. The PR looks really solid. Most of the feedback is just some things to make sure were sticking to the conventions we've set. This makes me realize that this stuff probably isnt documented well enough. We can fix that 😄

**Review by daniel-zamora — APPROVED** [review]

**Review by trevorlang — APPROVED** [review]

File: static/src/js/components/AccessMethod/Select.scss [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+  background-color: $color__white;
+  margin-bottom: 0.5rem;
+  margin-left: 1rem;
+  &:hover {
+    background-color: $color__black--100;
+  }
+}
+
+.SelectIcon {
+  float: right;
```
[the remark this answers is not in the document]
Comment (rushgeo) [reply static/src/js/components/AccessMethod/Select.scss]: I took the floats out in [7608b50](https://github.com/nasa/earthdata-search/pull/1726/commits/7608b509734dcfffddcce1d35f6aa8da54fbf95c) and replaced them with a combination of relative and absolute positioning. If we'd be better off using flexbox, we can switch to that when we revisit this component in the upcoming SWODLR work.
